keylime/tenant.py

```diff
@@ -505,8 +503,21 @@ def do_cvdelete(self):
         if response.status_code != 200:
             logger.error("Delete command response: %d Unexpected response from Cloud Verifier."%response.status_code)
             common.log_http_response(logger,logging.ERROR,response.json())
+            return
+        
+        deleted = False
+        for _ in range(12):
+            response = tornado_requests.request("GET", "http://%s:%s/instances/%s"%(self.cloudverifier_ip,self.cloudverifier_port,self.node_uuid),context=self.context)
+            if response.status_code == 404:
```

Comment:
I think you should be looking for a `200` here?

    DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000

Reply:
actually that call is to get the status of the node, so i want i to show a 404 to confirm that the node's been deleted from the database.  do you see any logs on the verifier that show that the status request is being serviced?

Reply:
actually the verifier doesn't log 404s i think.

Reply:
https://github.com/keylime/python-keylime/blob/master/keylime/cloud_verifier_tornado.py#L114

Reply:
It looks like the delete is actually failing somehow:

```
[root@localhost ~]# keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -c delete
Using config file /etc/keylime.conf
2019-03-11 17:09:29.664 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:09:29.664 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:09:34.739 - keylime.tenant - ERROR - Timed out waiting for delete of node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 to complete at CV
```

verifier:

```
2019-03-11 17:09:29.685 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
```

I then make up a UUID:

```
[root@localhost ~]# keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81CXXXXX -c delete
Using config file /etc/keylime.conf
2019-03-11 17:09:45.092 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:09:45.092 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:09:45.106 - keylime.tenant - ERROR - Delete command response: 404 Unexpected response from Cloud Verifier.
2019-03-11 17:09:45.106 - keylime.tenant - ERROR - Response code 404: instance id not found
```

```
2019-03-11 17:09:45.104 - keylime.cloudverifier - INFO - DELETE returning 404 response. instance id: D432FBB3-D2F1-4A97-9EF7-75BD81CXXXXX not found.
```

Reply:
The message "Timed out waiting for delete of node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 to complete at CV" actually comes from the status check.  the made up uuid works as expected.  can you try doing an add after the the failed delete?

Reply:
I think we might be able to ignore the above, this is from when a failed tenant quote state exists.

Reply:
some more details:

get the node to fail tenant quote:

```
2019-03-11 17:34:34.458 - keylime.cloudverifier - WARNING - Instance D432FBB3-D2F1-4A97-9EF7-75BD81C00000 has failed tenant quote.  stopping polling
```

Now try to add it again:

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -f somefile --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000
Using config file /etc/keylime.conf
2019-03-11 17:37:09.147 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:37:09.147 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:37:09.152 - keylime.tenant - INFO - TPM PCR Mask from policy is 0x408000
2019-03-11 17:37:09.152 - keylime.tenant - INFO - vTPM PCR Mask from policy is 0x808000
2019-03-11 17:37:09.153 - keylime.tenant - DEBUG - b64_v:g6vtZCL20tNPCugfz4hGODWAf+r7yUgDzJpCmI8Se3g=
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 745, in main
    mytenant.do_cv()
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 481, in do_cv
    raise Exception("Node %s already existed at CV.  Please use delete or update."%self.node_uuid)
Exception: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update.
2019-03-11 17:37:09.162 - keylime.tenant - ERROR - None
2019-03-11 17:37:09.162 - keylime.tenant - ERROR - Error: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update. 
```

So the CV believes it exists (which is right).

So now we delete:

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -f somefile --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000
Using config file /etc/keylime.conf
2019-03-11 17:37:09.147 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:37:09.147 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:37:09.152 - keylime.tenant - INFO - TPM PCR Mask from policy is 0x408000
2019-03-11 17:37:09.152 - keylime.tenant - INFO - vTPM PCR Mask from policy is 0x808000
2019-03-11 17:37:09.153 - keylime.tenant - DEBUG - b64_v:g6vtZCL20tNPCugfz4hGODWAf+r7yUgDzJpCmI8Se3g=
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 745, in main
    mytenant.do_cv()
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 481, in do_cv
    raise Exception("Node %s already existed at CV.  Please use delete or update."%self.node_uuid)
Exception: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update.
2019-03-11 17:37:09.162 - keylime.tenant - ERROR - None
2019-03-11 17:37:09.162 - keylime.tenant - ERROR - Error: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update. 
```

I can however delete when the node has not failed quote:

```
2019-03-11 17:39:40.212 - keylime.registrar_client - INFO - Registration activated for node D432FBB3-D2F1-4A97-9EF7-75BD81C00000.
```

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -c delete
Using config file /etc/keylime.conf
2019-03-11 17:40:02.940 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:40:02.940 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:40:02.975 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from the CV
```

Reply:
Some more details.

Get the node into a failed quote state:

```
2019-03-11 17:46:28.144 - keylime.cloudverifier - WARNING - Instance D432FBB3-D2F1-4A97-9EF7-75BD81C00000 has failed tenant quote.  stopping polling
```

Try to delete the node:

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -c delete
Using config file /etc/keylime.conf
2019-03-11 17:50:00.290 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:50:00.290 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:50:05.409 - keylime.tenant - ERROR - Timed out waiting for delete of node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 to complete at CV
```

If its a registered node not in a failed quote state, it deletes ok:

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -c delete
Using config file /etc/keylime.conf
2019-03-11 17:48:25.801 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:48:25.801 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:48:25.844 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from the CV
```

The verifier is not giving us much more than the 200 each time:

```
2019-03-11 17:51:46.945 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
```

Reply:
yeah, the verifier just marks the node for deletion and returns 200.  I thought that this code would just immediately delete it if it was already failed
https://github.com/keylime/python-keylime/blob/7b0a5b4c57e0efd7add2cac88a7ad0eca95e9a90/keylime/cloud_verifier_tornado.py#L151-L154

I'll debug a little and see what's going on

Reply:
hrm, i wasn't able to reproduce this one.  when it is in a failed state it does delete for me.